diff --git a/wifi/wifi.c b/wifi/wifi.c
index d62b30f..ea0e6f9 100644
--- a/wifi/wifi.c
+++ b/wifi/wifi.c
@@ -574,6 +574,12 @@ int wifi_start_supplicant(int p2p_supported)
         serial = pi->serial;
     }
 #endif
+    /* The ar6k driver needs the interface up in order to scan! */
+    if (!strncmp(DRIVER_MODULE_NAME, "ar6000", 6)) {
+        ifc_init();
+        ifc_up("wlan0");
+        sleep(2);
+    }
     property_get("wifi.interface", primary_iface, WIFI_TEST_INTERFACE);
 
     property_set("ctl.start", supplicant_name);
